File: code/java/ch19/learn/Example02BfsVsDfs.java
package ch19.learn;

import java.util.*;

/**
 * Example 02: BFS vs DFS
 * ========================
 * Chapter 19: Graphs I — Exploring Networks
 *
 * Demonstrates:
 *   Part 1: BFS traversal step by step
 *   Part 2: DFS traversal step by step
 *   Part 3: BFS finds shortest paths, DFS does not
 */
public class Example02BfsVsDfs {

    public static void main(String[] args) {
        // Graph:
        //   0 --- 1
        //   |     |
        //   2 --- 3
        //         |
        //         4
        int n = 5;
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());
        int[][] edges = {{0,1}, {0,2}, {1,3}, {2,3}, {3,4}};
        for (int[] e : edges) {
            adj.get(e[0]).add(e[1]);
            adj.get(e[1]).add(e[0]);
        }

        // ── Part 1: BFS ───────────────────────────────────
        System.out.println("=== Part 1: BFS from node 0 ===");
        boolean[] visitedBfs = new boolean[n];
        visitedBfs[0] = true;
        Queue<Integer> queue = new LinkedList<>();
        queue.add(0);
        List<Integer> bfsOrder = new ArrayList<>();

        while (!queue.isEmpty()) {
            int node = queue.poll();
            bfsOrder.add(node);
            List<Integer> nbrs = new ArrayList<>(adj.get(node));
            Collections.sort(nbrs);
            for (int nb : nbrs) {
                if (!visitedBfs[nb]) {
                    visitedBfs[nb] = true;
                    queue.add(nb);
                }
            }
        }
        System.out.println("  BFS order: " + bfsOrder);

        // ── Part 2: DFS ───────────────────────────────────
        System.out.println("\n=== Part 2: DFS from node 0 ===");
        boolean[] visitedDfs = new boolean[n];
        List<Integer> dfsOrder = new ArrayList<>();
        dfs(adj, 0, visitedDfs, dfsOrder);
        System.out.println("  DFS order: " + dfsOrder);

        // ── Part 3: Shortest Paths ─────────────────────────
        System.out.println("\n=== Part 3: Shortest Paths (BFS vs DFS) ===");
        // Smaller graph: 0-1, 0-2, 1-3, 2-3
        int m = 4;
        List<List<Integer>> adj2 = new ArrayList<>();
        for (int i = 0; i < m; i++) adj2.add(new ArrayList<>());
        adj2.get(0).add(1); adj2.get(1).add(0);
        adj2.get(0).add(2); adj2.get(2).add(0);
        adj2.get(1).add(3); adj2.get(3).add(1);
        adj2.get(2).add(3); adj2.get(3).add(2);

        // BFS distances
        int[] distBfs = new int[m];
        Arrays.fill(distBfs, -1);
        distBfs[0] = 0;
        Queue<Integer> q2 = new LinkedList<>();
        q2.add(0);
        while (!q2.isEmpty()) {
            int node = q2.poll();
            for (int nb : adj2.get(node)) {
                if (distBfs[nb] == -1) {
                    distBfs[nb] = distBfs[node] + 1;
                    q2.add(nb);
                }
            }
        }
        System.out.println("  BFS distances from 0: " + Arrays.toString(distBfs));
        System.out.println("  BFS is CORRECT for shortest paths!");
        System.out.println("  DFS would give wrong distances (e.g., dist[2]=3 instead of 1)");
    }

    static void dfs(List<List<Integer>> adj, int node,
                     boolean[] visited, List<Integer> order) {
        visited[node] = true;
        order.add(node);
        List<Integer> nbrs = new ArrayList<>(adj.get(node));
        Collections.sort(nbrs);
        for (int nb : nbrs) {
            if (!visited[nb]) {
                dfs(adj, nb, visited, order);
            }
        }
    }
}

File: code/cpp/ch19/learn/example_02_bfs_vs_dfs.cpp
/*
 * Example 02: BFS vs DFS
 * ========================
 * Chapter 19: Graphs I — Exploring Networks
 *
 * Demonstrates:
 *   Part 1: BFS traversal
 *   Part 2: DFS traversal
 *   Part 3: BFS finds shortest paths, DFS does not
 */

#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

void dfsHelper(vector<vector<int>>& adj, int node,
               vector<bool>& visited, vector<int>& order) {
    visited[node] = true;
    order.push_back(node);
    vector<int> nbrs = adj[node];
    sort(nbrs.begin(), nbrs.end());
    for (int nb : nbrs) {
        if (!visited[nb]) dfsHelper(adj, nb, visited, order);
    }
}

int main() {
    int n = 5;
    vector<vector<int>> adj(n);
    int edges[][2] = {{0,1},{0,2},{1,3},{2,3},{3,4}};
    for (auto& e : edges) {
        adj[e[0]].push_back(e[1]);
        adj[e[1]].push_back(e[0]);
    }

    // ── Part 1: BFS ───────────────────────────────
    cout << "=== Part 1: BFS from node 0 ===" << endl;
    vector<bool> visitedBfs(n, false);
    visitedBfs[0] = true;
    queue<int> q;
    q.push(0);
    vector<int> bfsOrder;
    while (!q.empty()) {
        int node = q.front(); q.pop();
        bfsOrder.push_back(node);
        vector<int> nbrs = adj[node];
        sort(nbrs.begin(), nbrs.end());
        for (int nb : nbrs) {
            if (!visitedBfs[nb]) {
                visitedBfs[nb] = true;
                q.push(nb);
            }
        }
    }
    cout << "  BFS order: ";
    for (int x : bfsOrder) cout << x << " ";
    cout << endl;

    // ── Part 2: DFS ───────────────────────────────
    cout << "\n=== Part 2: DFS from node 0 ===" << endl;
    vector<bool> visitedDfs(n, false);
    vector<int> dfsOrder;
    dfsHelper(adj, 0, visitedDfs, dfsOrder);
    cout << "  DFS order: ";
    for (int x : dfsOrder) cout << x << " ";
    cout << endl;

    // ── Part 3: Shortest Paths ────────────────────
    cout << "\n=== Part 3: Shortest Paths ===" << endl;
    // Smaller graph: 0-1, 0-2, 1-3, 2-3
    int m = 4;
    vector<vector<int>> adj2(m);
    adj2[0].push_back(1); adj2[1].push_back(0);
    adj2[0].push_back(2); adj2[2].push_back(0);
    adj2[1].push_back(3); adj2[3].push_back(1);
    adj2[2].push_back(3); adj2[3].push_back(2);

    vector<int> dist(m, -1);
    dist[0] = 0;
    queue<int> q2;
    q2.push(0);
    while (!q2.empty()) {
        int node = q2.front(); q2.pop();
        for (int nb : adj2[node]) {
            if (dist[nb] == -1) {
                dist[nb] = dist[node] + 1;
                q2.push(nb);
            }
        }
    }
    cout << "  BFS distances from 0: ";
    for (int d : dist) cout << d << " ";
    cout << endl;
    cout << "  BFS is CORRECT for shortest paths!" << endl;

    return 0;
}
